sql/core/src/test/scala/org/apache/spark/sql/ProductAggSuite.scala

```diff
@@ -0,0 +1,107 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.expressions.Window
+import org.apache.spark.sql.functions.{ col, lit, product }
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.{ByteType, DoubleType, FloatType, IntegerType, ShortType}
+
+
+class ProductAggSuite extends QueryTest with SharedSparkSession {
+  // Sequence of integers small enough that factorial is representable exactly as DoubleType:
+  private lazy val data16 = spark.range(1, 17).toDF("x")
+
+  private lazy val factorials = (1 to 16).scanLeft(1L) { case (f, x) => f * x }
+
+  test("bare factorial") {
+    implicit val enc = Encoders.scalaDouble
+
+    val prod = data16.agg(product(col("x"))).as[Double].head
+    val expected = (1L to 16L).reduce { _ * _ }.toDouble
+
+    assert(prod === expected)
+    assert(prod === factorials(16))
+  }
+
+  test("type flexibility") {
+    val bytes16 = spark.createDataset((1 to 16).map { _.toByte })(Encoders.scalaByte).toDF("x")
+
+    val variants = Map(
+      "int8" -> ByteType, "int16" -> ShortType, "int32" -> IntegerType,
+      "float32" -> FloatType, "float64" -> DoubleType)
+
+    val prods = variants.foldLeft(bytes16) { case (df, (id, typ)) =>
+      df.withColumn(id, df.col("x").cast(typ))
+    }.agg(
+      lit(1) as "dummy",
+      variants.keys.toSeq.map { id => product(col(id)) as id } : _*)
+
+    variants.keys.foreach { typ =>
+      val prod = prods.select(typ).as[Double](Encoders.scalaDouble).head
+      assert(prod === factorials(16))
+    }
+  }
+
+  test("windowed factorials") {
+    implicit val enc = Encoders.product[(Long, Double)]
+    val win = Window.partitionBy(lit(1)).orderBy("x")
+
+    val prodFactorials = data16.withColumn("f", product(col("x")).over(win))
+
+    val prodMap = prodFactorials.as[(Long, Double)].collect.toMap
+
+    assert(prodMap.size === 16)
+
+    assert(prodMap(1) === 1.0)
+    assert(prodMap(2) === 2.0)
+    assert(prodMap(3) === 6.0)
+    assert(prodMap(4) === 24.0)
+    assert(prodMap(5) === 120.0)
+
+    factorials.zipWithIndex.drop(1).foreach { case (expected, idx) =>
+      assert(prodMap(idx) === expected)
+    }
+  }
+
+  test("grouped factorials") {
+    implicit val enc = Encoders.scalaDouble
+
+    val grouped = data16.groupBy((col("x") % 3) as "mod3")
+      .agg(
+        product(col("x")) as "product",
+        product(col("x") * 0.5) as "product_scaled",
+        product(col("x") * -1.0) as "product_minus")
+      .orderBy("mod3")
+
+    def col2seq(s: String): Seq[Double] =
+      grouped.select(s).as[Double].collect.toSeq
+
+    val expectedBase = Seq((3 * 6 * 9 * 12 * 15),
+                           (1 * 4 * 7 * 10 * 13 * 16),
+                           (2 * 5 * 8 * 11 * 14))
```

Comment:
```suggestion
    val expectedBase = Seq(
      (3 * 6 * 9 * 12 * 15),
      (1 * 4 * 7 * 10 * 13 * 16),
      (2 * 5 * 8 * 11 * 14))
```

What about this version?  There's actually a style guide here to refer (https://github.com/databricks/scala-style-guide#spacing-and-indentation)